LayoutTest/TestCase/LYTLayoutFailingTestSnapshotRecorder.m

```diff
@@ -0,0 +1,233 @@
+}
+
+@interface LYTLayoutFailingTestSnapshotRecorder ()
+
+@property (nonatomic) Class invocationClass;
+@property (nonatomic) NSMutableSet *failingTestsSnapshotFolders;
+
+@end
+
+@implementation LYTLayoutFailingTestSnapshotRecorder
+
++ (instancetype)sharedInstance {
+    static LYTLayoutFailingTestSnapshotRecorder *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[LYTLayoutFailingTestSnapshotRecorder alloc] init];
+    });
+    return sharedInstance;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.failingTestsSnapshotFolders = [NSMutableSet set];
+        XCTestObservationCenter *observationCenter = [XCTestObservationCenter sharedTestObservationCenter];
+        [observationCenter addTestObserver:self];
+    }
+    return self;
+}
+
+- (void)testBundleDidFinish:(NSBundle *)testBundle {
+    if (self.failingTestsSnapshotFolders.count > 0) {
+        SimpleLog(@"\nSnapshots of failing tests can be found in:\n");
+        for (NSString *path in self.failingTestsSnapshotFolders) {
+            SimpleLog(@"%@\n", path);
+        }
+        [self.failingTestsSnapshotFolders removeAllObjects];
+    }
+}
+
+- (void)testCase:(XCTestCase *)testCase didFailWithDescription:(NSString *)description inFile:(nullable NSString *)filePath atLine:(NSUInteger)lineNumber {
+    if ([testCase isKindOfClass:[LYTLayoutTestCase class]]) {
+        NSString *pathForSnapshots = [self commonRootPathForInvocationClass:[testCase class]];
+        pathForSnapshots = [pathForSnapshots stringByAppendingString:@"/index.html"];
+        [self.failingTestsSnapshotFolders addObject:pathForSnapshots];
+    }
+}
+
+- (void)startNewLogForClass:(Class)invocationClass {
+    self.invocationClass = invocationClass;
+    [self deleteCurrentFailingSnapshots];
+    [self createIndexHTMLFile];
+}
+
+- (void)saveImageOfView:(UIView *)view withData:(NSDictionary *)data fromInvocation:(NSInvocation *)invocation failureDescription:(NSString *)failureDescription {
+    NSString *imagePath = [self pathForImageWithWidth:view.frame.size.width height:view.frame.size.height data:data invocation:invocation];
+    if ([self shouldSaveImageOfViewatPath:imagePath withInvocation:invocation]) {
+        [self createDirectoryForInvocationIfNeeded:invocation];
+        UIImage *viewImage = [self renderLayer:view.layer];
+        [UIImagePNGRepresentation(viewImage) writeToFile:imagePath atomically:YES];
+        [self appendViewWithWidth:view.frame.size.width
+                           height:view.frame.size.height
+                         viewData:data
+      toLogWithFailureDescription:failureDescription
+                   withInvocation:invocation];
+    }
+}
+
+- (BOOL)shouldSaveImageOfViewatPath:(NSString *)imagePath withInvocation:(NSInvocation *)invocation {
+    if ([[NSFileManager defaultManager] fileExistsAtPath:imagePath]) {
+        return NO;
```

Comment:
Added incase there is a view/data combo failure that is repeated. Unlikely to happen I think but should help save some time creating the image and writing it again.